src/main/java/org/jenkinsci/plugins/p4/scm/BranchesScmSource.java

```diff
@@ -0,0 +1,65 @@
+package org.jenkinsci.plugins.p4.scm;
+
+import com.perforce.p4java.core.file.IFileSpec;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.model.TaskListener;
+import org.jenkinsci.plugins.p4.browsers.P4Browser;
+import org.jenkinsci.plugins.p4.client.ConnectionHelper;
+import org.jenkinsci.plugins.p4.workspace.ManualWorkspaceImpl;
+import org.jenkinsci.plugins.p4.workspace.Workspace;
+import org.jenkinsci.plugins.p4.workspace.WorkspaceSpec;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+public class BranchesScmSource extends AbstractP4ScmSource {
+
+	@DataBoundConstructor
+	public BranchesScmSource(String id, String credential, String includes, String charset, String format, P4Browser browser) {
+		super(id, credential, includes, charset, format, browser);
+	}
+
+	public List<P4Head> getHeads(@NonNull TaskListener listener) throws Exception {
+
+		List<String> paths = getIncludePaths();
+		HashSet<P4Head> list = new HashSet<P4Head>();
+
+		ConnectionHelper p4 = new ConnectionHelper(credential, listener);
```

Comment:
You should consider making this class Closeable so that you can use try with resources